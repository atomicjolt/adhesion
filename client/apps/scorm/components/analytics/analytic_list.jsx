import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import _ from 'lodash';
import AnalyticRow from './analytic_row';
import { switchView } from '../../actions/analytics';

export class AnalyticList extends React.Component {

  static propTypes = {
    switchView: PropTypes.func.isRequired,
    tableData: PropTypes.array,
    view: PropTypes.string.isRequired,
  }

  constructor() {
    super();
    this.state = {
      analyticsName: '',
      parents: {},
      sortAsc: true,
      sortBy: 'name',
    };
  }
  componentWillMount() {
    this.setAnalyticsName(this.props.view);
    this.setParent(this.props.tableData);
  }

  componentWillReceiveProps(nextProps) {
    this.setAnalyticsName(nextProps.view);
    this.setParent(nextProps.tableData);
  }

  setAnalyticsName(view) {
    if (view === 'course') {
      this.setState({ analyticsName: 'Student' });
    } else {
      this.setState({ analyticsName: 'Activity' });
    }
  }

  setParent(tableData) {
    let parents = _.map(tableData, (reg) => {
      const data = {};
      if (_.size(reg.childrenIds) > 0) {
        data.id = reg.id;
        data.show = true;
        data.childrenIds = reg.childrenIds;
      }
      return data;
    });
    parents = _.remove(parents, {});
    this.setState({ parents });
  }

  tableRowClicked(viewId) {
    const parent = _.find(this.state.parents, { id: viewId });
    if (parent) {
      const show = !parent.show;
      const parents = this.hideChildren(this.state.parents, viewId, show);
      this.setState({ parents });
    }
  }

  hideChildren(parents, id, show) {
    const data = _.find(parents, { id });
    if (data) {
      data.show = show;
      if (data.childrenIds) {
        _.each(data.childrenIds, (childId) => {
          this.hideChildren(parents, childId, show);
        });
      }
    }
    return parents;
  }

  sort(col) {
    this.setState(prevState => (
      {
        sortAsc: !prevState.sortAsc,
        sortBy: col,
      }
    ));
  }

  sortIcon(key) {
    if (this.state.sortBy === key && this.props.view === 'course') {
      if (this.state.sortAsc) return 'keyboard_arrow_up';
      return 'keyboard_arrow_down';
    }
    return null;
  }

  data() {
    const asc = this.state.sortAsc ? 'asc' : 'desc';
    const data = this.props.view === 'course' ?
      _.orderBy(this.props.tableData, [this.state.sortBy], [asc]) :
      this.props.tableData;
    return (
      _.map(data, (reg, key) => {
        const parent = _.find(this.state.parents, { id: reg.parentId });
        return (<AnalyticRow
          key={key}
          id={parseInt(reg.id, 10)}
          name={reg.name}
          passed={reg.passed}
          score={reg.score}
          time={reg.time}
          show={parent ? parent.show : true}
          isParent={_.size(reg.childrenIds) > 0}
          depth={reg.depth}
          tableRowClicked={viewId => this.tableRowClicked(viewId)}
        />);
      })
    );
  }

  tableHeaders() {
    const headers = [
      { name: 'name', label: this.state.analyticsName },
      { name: 'passed', label: 'Passed' },
      { name: 'score', label: 'AVG Score' },
      { name: 'time', label: 'Total Time(mins)' },
    ];

    return (
      _.map(headers, header => (
        <th
          key={header.name}
          onClick={() => { this.sort(header.name); }}
        >
          {header.label}
          <i className="material-icons">
            {this.sortIcon(header.name)}
          </i>
        </th>
      ))
    );
  }

  render() {

    return (
      <table className="c-aa-table">
        <thead>
          <tr>
            { this.tableHeaders() }
          </tr>
        </thead>
        <tbody>
          { this.data() }
        </tbody>
      </table>
    );
  }
}

const select = state => ({
  view: state.analytics.view,
});

export default connect(select, { switchView })(AnalyticList);
