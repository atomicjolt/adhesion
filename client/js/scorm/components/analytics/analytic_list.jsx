import React              from 'react';
import { connect }        from 'react-redux';
import _                  from 'lodash';
import AnalyticRow        from './analytic_row';
import { switchView }     from '../../actions/analytics';

export class AnalyticList extends React.Component {

  static propTypes = {
    switchView: React.PropTypes.func.isRequired,
    tableData: React.PropTypes.array,
    view: React.PropTypes.string,
  }

  constructor() {
    super();
    this.state = {
      analyticsName: '',
    };
  }
  componentWillMount() {
    this.setAnalyticsName(this.props.view);
  }

  componentWillReceiveProps(nextProps) {
    this.setAnalyticsName(nextProps.view);
  }

  setAnalyticsName(view) {
    if (view === 'course') {
      this.setState({ analyticsName: 'Student' });
    } else {
      this.setState({ analyticsName: 'Activity' });
    }
  }

  switchTable(viewId) {
    if(viewId) {
      this.props.switchView('student', viewId);
    }
  }

  render() {
    return (
      <table className="c-aa-table">
        <thead>
          <tr>
            <th>{this.state.analyticsName}</th>
            <th>Passed</th>
            <th>AVG Score</th>
            <th>AVG Time(mins)</th>
          </tr>
        </thead>
        <tbody>
          {
            _.map(this.props.tableData, (reg, key) => (
              <AnalyticRow
                key={key}
                id={reg.id}
                name={reg.name}
                passed={reg.passed}
                score={reg.score}
                time={reg.time}
                switchTable={this.switchTable.bind(this)}
              />
            ))
          }
        </tbody>
      </table>
    );
  }
}

const select = state => ({
  view: state.analytics.view,
});

export default connect(select, { switchView })(AnalyticList);
