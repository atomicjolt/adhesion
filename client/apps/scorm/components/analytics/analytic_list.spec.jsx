import React from 'react';
import { shallow } from 'enzyme';
import { AnalyticList } from './analytic_list';

describe('Scorm Analytics AnalyticList', () => {

  let result;
  let props;

  describe('Scorm Analytics AnalyticList', () => {
    beforeEach(() => {
      props = {
        tableData: [],
        view: 'course',
        switchView: () => {},
      };
      result = shallow(<AnalyticList {...props} />);
    });

    it('renders', () => {
      expect(result).toMatchSnapshot();
    });

    it('renders the analytics list with the correct header value', () => {
      const table = result.find('.c-aa-table th').first();
      expect(table.text()).toContain('Student');
    });

    it('renders the analytics list with the correct header value', () => {
      props.view = '';
      result = shallow(<AnalyticList {...props} />);
      const table = result.find('.c-aa-table th').first();
      expect(table.text()).toContain('Activity');
    });

    it('renders the table rows with the correct amount of rows', () => {
      const trs = result.find('.c-aa-table tr');
      expect(trs.length).toBe(1);
    });
  });
});
