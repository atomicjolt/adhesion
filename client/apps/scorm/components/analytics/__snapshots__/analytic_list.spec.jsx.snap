// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scorm Analytics AnalyticList Scorm Analytics AnalyticList renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AnalyticList
    switchView={[Function]}
    tableData={Array []}
    view="course"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <thead>
          <tr>
            <th
              onClick={[Function]}
            >
              Student
              <i
                className="material-icons"
              >
                keyboard_arrow_up
              </i>
            </th>
            <th
              onClick={[Function]}
            >
              Passed
              <i
                className="material-icons"
              />
            </th>
            <th
              onClick={[Function]}
            >
              AVG Score
              <i
                className="material-icons"
              />
            </th>
            <th
              onClick={[Function]}
            >
              Total Time(mins)
              <i
                className="material-icons"
              />
            </th>
          </tr>
        </thead>,
        <tbody />,
      ],
      "className": "c-aa-table",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <tr>
            <th
              onClick={[Function]}
            >
              Student
              <i
                className="material-icons"
              >
                keyboard_arrow_up
              </i>
            </th>
            <th
              onClick={[Function]}
            >
              Passed
              <i
                className="material-icons"
              />
            </th>
            <th
              onClick={[Function]}
            >
              AVG Score
              <i
                className="material-icons"
              />
            </th>
            <th
              onClick={[Function]}
            >
              Total Time(mins)
              <i
                className="material-icons"
              />
            </th>
          </tr>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <th
                onClick={[Function]}
              >
                Student
                <i
                  className="material-icons"
                >
                  keyboard_arrow_up
                </i>
              </th>,
              <th
                onClick={[Function]}
              >
                Passed
                <i
                  className="material-icons"
                />
              </th>,
              <th
                onClick={[Function]}
              >
                AVG Score
                <i
                  className="material-icons"
                />
              </th>,
              <th
                onClick={[Function]}
              >
                Total Time(mins)
                <i
                  className="material-icons"
                />
              </th>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Student",
                  <i
                    className="material-icons"
                  >
                    keyboard_arrow_up
                  </i>,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Student",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "keyboard_arrow_up",
                    "className": "material-icons",
                  },
                  "ref": null,
                  "rendered": "keyboard_arrow_up",
                  "type": "i",
                },
              ],
              "type": "th",
            },
            Object {
              "instance": null,
              "key": "passed",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Passed",
                  <i
                    className="material-icons"
                  />,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Passed",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": null,
                    "className": "material-icons",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
              ],
              "type": "th",
            },
            Object {
              "instance": null,
              "key": "score",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "AVG Score",
                  <i
                    className="material-icons"
                  />,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "AVG Score",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": null,
                    "className": "material-icons",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
              ],
              "type": "th",
            },
            Object {
              "instance": null,
              "key": "time",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Total Time(mins)",
                  <i
                    className="material-icons"
                  />,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Total Time(mins)",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": null,
                    "className": "material-icons",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
              ],
              "type": "th",
            },
          ],
          "type": "tr",
        },
        "type": "thead",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": "tbody",
      },
    ],
    "type": "table",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
              <th
                onClick={[Function]}
              >
                Student
                <i
                  className="material-icons"
                >
                  keyboard_arrow_up
                </i>
              </th>
              <th
                onClick={[Function]}
              >
                Passed
                <i
                  className="material-icons"
                />
              </th>
              <th
                onClick={[Function]}
              >
                AVG Score
                <i
                  className="material-icons"
                />
              </th>
              <th
                onClick={[Function]}
              >
                Total Time(mins)
                <i
                  className="material-icons"
                />
              </th>
            </tr>
          </thead>,
          <tbody />,
        ],
        "className": "c-aa-table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th
                onClick={[Function]}
              >
                Student
                <i
                  className="material-icons"
                >
                  keyboard_arrow_up
                </i>
              </th>
              <th
                onClick={[Function]}
              >
                Passed
                <i
                  className="material-icons"
                />
              </th>
              <th
                onClick={[Function]}
              >
                AVG Score
                <i
                  className="material-icons"
                />
              </th>
              <th
                onClick={[Function]}
              >
                Total Time(mins)
                <i
                  className="material-icons"
                />
              </th>
            </tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th
                  onClick={[Function]}
                >
                  Student
                  <i
                    className="material-icons"
                  >
                    keyboard_arrow_up
                  </i>
                </th>,
                <th
                  onClick={[Function]}
                >
                  Passed
                  <i
                    className="material-icons"
                  />
                </th>,
                <th
                  onClick={[Function]}
                >
                  AVG Score
                  <i
                    className="material-icons"
                  />
                </th>,
                <th
                  onClick={[Function]}
                >
                  Total Time(mins)
                  <i
                    className="material-icons"
                  />
                </th>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "name",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Student",
                    <i
                      className="material-icons"
                    >
                      keyboard_arrow_up
                    </i>,
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "Student",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "keyboard_arrow_up",
                      "className": "material-icons",
                    },
                    "ref": null,
                    "rendered": "keyboard_arrow_up",
                    "type": "i",
                  },
                ],
                "type": "th",
              },
              Object {
                "instance": null,
                "key": "passed",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Passed",
                    <i
                      className="material-icons"
                    />,
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "Passed",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": null,
                      "className": "material-icons",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "i",
                  },
                ],
                "type": "th",
              },
              Object {
                "instance": null,
                "key": "score",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "AVG Score",
                    <i
                      className="material-icons"
                    />,
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "AVG Score",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": null,
                      "className": "material-icons",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "i",
                  },
                ],
                "type": "th",
              },
              Object {
                "instance": null,
                "key": "time",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Total Time(mins)",
                    <i
                      className="material-icons"
                    />,
                  ],
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "Total Time(mins)",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": null,
                      "className": "material-icons",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "i",
                  },
                ],
                "type": "th",
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": "tbody",
        },
      ],
      "type": "table",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
